frame/crowdloan-rewards/src/lib.rs

```diff
@@ -0,0 +1,325 @@
+				},
+			);
+			TotalRewards::<T>::set(total_rewards);
+			TotalContributors::<T>::set(total_contributors as u32);
+		}
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		NotInitialized,
+		AlreadyInitialized,
+		InvalidProof,
+		InvalidClaim,
+		NotAContributor,
+		NothingToClaim,
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		#[allow(missing_docs)]
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type Currency: Mutate<Self::AccountId, Balance = Self::Balance>;
+		type Balance: Default
+			+ Parameter
+			+ Codec
+			+ Copy
+			+ Ord
+			+ CheckedAdd
+			+ CheckedSub
+			+ CheckedMul
+			+ AtLeast32BitUnsigned
+			+ MaybeSerializeDeserialize
+			+ Zero;
+		type AdminOrigin: EnsureOrigin<Self::Origin>;
+		type Convert: Convert<Self::BlockNumber, Self::Balance>;
+		type RelayChainAccountId: Parameter + MaybeSerializeDeserialize + Into<AccountId32> + Ord;
+		#[pallet::constant]
+		type InitialPayment: Get<Perbill>;
+		#[pallet::constant]
+		type VestingPartition: Get<Self::BlockNumber>;
+		#[pallet::constant]
+		type Prefix: Get<&'static [u8]>;
+	}
+
+	#[pallet::storage]
+	pub type Rewards<T: Config> =
+		StorageMap<_, Blake2_128Concat, RemoteAccountOf<T>, RewardOf<T>, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn total_rewards)]
+	pub type TotalRewards<T: Config> = StorageValue<_, T::Balance, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn claimed_rewards)]
+	pub type ClaimedRewards<T: Config> = StorageValue<_, T::Balance, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn total_contributors)]
+	pub type TotalContributors<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn vesting_block_start)]
+	pub type VestingBlockStart<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(10_000)]
+		pub fn initialize(origin: OriginFor<T>) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+			ensure!(!VestingBlockStart::<T>::exists(), Error::<T>::AlreadyInitialized);
+			let current_block = frame_system::Pallet::<T>::block_number();
+			VestingBlockStart::<T>::set(current_block);
+			Ok(())
+		}
+
+		#[pallet::weight(10_000)]
+		pub fn claim_from_ethereum(
+			origin: OriginFor<T>,
+			reward_account: T::AccountId,
+			ethereum_address: EthereumAddress,
+			proof: EcdsaSignature,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			let data = reward_account.using_encoded(to_ascii_hex);
+			let signer = ethereum_recover(T::Prefix::get(), &proof, &data)
+				.ok_or(Error::<T>::InvalidProof)?;
+			// NOTE(hussein-aitlahcen): ask for the address even if we can extract it
+			// from the signature as we want to differentiate invalid proof from non-existant
+			// contributors
+			ensure!(signer == ethereum_address, Error::<T>::InvalidProof);
```

Comment:
invalid proof if data is not eth signed versioned msg. if sinature is valid, but if signer not in list, why would on chain code care? it is up to claimer to recheck. so it seems that such distinctions is up to offchain code?

Reply:
just `what can be done offchain should be done off chain`?

Reply:
I added it to make the claiming mechanism having a consistent behavior for both eth/relay claims. With the multisignature the proof will be considered invalid but since we extract the account from the eth one,  it will result   in an "NotAContributor" error.

Reply:
if feature has internal assymetry and public api has some kind of assymetry, why would implementation in terms of errors to be symmetrical? sounds like OOP:)

Reply:
I don't really like this extra "useless" information but as a user it is perhaps better to know if you failed to sign or signed a wrong message vs having a "you are not a contributor" answer.  The API asymmetry is present for the same reason, I feel like it's easier for a user to use a different extrinsic based on what type of contributor it was (relay/eth). I don't have a strong opinion on that, I can remove it :).

Reply:
error could be SignatureOfMessageNotFound in contributor list. Js client can than show list to user and he can match hist metamask to our list. doubt it ever will be backend problem as it should be solved by ui. or you clever to solve it properly by hands, than error is good enough for you.